### PR TITLE
Add set ID to Private Set Intersection

### PR DESCRIPTION
**Pull request by ipetr0v**
This change:
- Adds a set ID to each PSI request
- Re-enables PSI C++ client

Fixes https://github.com/project-oak/oak/issues/1537

### CODE REVIEW

File: examples/private_set_intersection/example.toml [hunk 1]
```diff
@@ -16,10 +16,11 @@ additional_args = [
 
 [clients]
 rust = { Cargo = { cargo_manifest = "examples/private_set_intersection/client/rust/Cargo.toml" }, additional_args = [
+  "--set-id=rust",
```
Comment (tiziano88) [inline examples/private_set_intersection/example.toml:19]: Perhaps add a comment somewhere around here explaining why these have different ids, and what they mean, etc.

File: examples/private_set_intersection/module/rust/src/lib.rs [hunk 2]
```diff
@@ -49,32 +49,46 @@ oak::entrypoint!(oak_main => |_in_channel| {
...
+    values: HashSet<String>,
     /// Number of contributed private sets.
     set_count: u64,
     /// The intersection is locked and new contributions are discarded.
     locked: bool,
 }
 
+#[derive(Default)]
+struct Node {
+    /// Map from set ID to `SetIntersection`.
```
Comment (tiziano88) [inline examples/private_set_intersection/module/rust/src/lib.rs:63]: Could you explain a bit more here? e.g. that this allows multiple sets of clients to compute their own intersections, also explain what the security characteristics are, in particular I think currently just specifying a different set id does not guarantee that data from different clients is kept separate, right? That would require some more sophisticated labelling.

File: examples/private_set_intersection/module/rust/src/lib.rs [hunk 3]
```diff
@@ -49,32 +49,46 @@ oak::entrypoint!(oak_main => |_in_channel| {
...
+    sets: HashMap<String, SetIntersection>,
+}
+
 impl PrivateSetIntersection for Node {
     fn submit_set(&mut self, req: SubmitSetRequest) -> grpc::Result<()> {
-        if self.locked {
+        let mut current_set = self
+            .sets
+            .entry(req.set_id)
+            .or_insert(SetIntersection::default());
```
Comment (tiziano88) [inline examples/private_set_intersection/module/rust/src/lib.rs:72]:
```suggestion
            .or_default();
```

File: examples/private_set_intersection/module/rust/src/lib.rs [hunk 4]
```diff
@@ -49,32 +49,46 @@ oak::entrypoint!(oak_main => |_in_channel| {
...
+        if current_set.set_count < SET_THRESHOLD {
+            let submitted_set = req.values.iter().cloned().collect::<HashSet<_>>();
+            current_set.values = if current_set.values.is_empty() {
+                submitted_set
+            } else {
+                current_set
+                    .values
+                    .intersection(&submitted_set)
+                    .cloned()
+                    .collect()
```
Comment (tiziano88) [inline examples/private_set_intersection/module/rust/src/lib.rs:90]: Where is the code that increments the sample count now?

If it's missing, how can it work?

**Review by tiziano88 — APPROVED** [review]

File: examples/private_set_intersection/module/rust/src/lib.rs (hunk 4 above)
Comment (ipetr0v) [reply examples/private_set_intersection/module/rust/src/lib.rs:90]: Done

File: examples/private_set_intersection/module/rust/src/lib.rs (hunk 3 above)
Comment (ipetr0v) [reply examples/private_set_intersection/module/rust/src/lib.rs:72]: Done

**Pull request merged**